Feature/cleanup history

poelzi opened the pull request:
Remove old history entries that don't have any played tracks.
When mixxx ist started and stopped without any tracks played, it
causes one history entry that is never used and bloats up the menu.

We simply delete all playlists (history entries) that don't have
any tracks associated.

poelzi committed:
Use qWarning on failed SQL queries

poelzi committed:
Remove empty history entries on startup

Remove old history entries that don't have any played tracks.
When mixxx ist started and stopped without any tracks played, it
causes one history entry that is never used and bloats up the menu.

We simply delete all playlists (history entries) that don't have
any tracks associated.

Be-ing commented:
Thanks for taking care of this. It has annoyed me when looking through my history. I'll test it soon.

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -414,6 +414,37 @@ void PlaylistDAO::removeHiddenTracks(const int playlistId) {
     emit(changed(playlistId));
 }
 
+void PlaylistDAO::removeEmptyPlaylists(const int hidden) {
```
Use HiddenType instead of int and omit 'const'

uklotzde commented on src/library/setlogfeature.cpp:
```diff
@@ -19,6 +19,8 @@ SetlogFeature::SetlogFeature(QObject* parent,
                                                    "mixxx.db.model.setlog",
                                                    true); //show all tracks
+    // we cleanup old history playlists that don't have any tracks
+    m_playlistDao.removeEmptyPlaylists(2);
```
2 == PLHT_SET_LOG (from HiddenType)

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -414,6 +414,37 @@ void PlaylistDAO::removeHiddenTracks(const int playlistId) {
 
+void PlaylistDAO::removeEmptyPlaylists(const int hidden) {
+    ScopedTransaction transaction(m_database);
+    // This query deletes all tracks marked as deleted and all
```
This comment does not match the code

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -414,6 +414,37 @@ void PlaylistDAO::removeHiddenTracks(const int playlistId) {
+    // phantom track_ids with no match in the library table
+    QString queryString = QString(
+        "SELECT id FROM Playlists "
+        "WHERE hidden=%1 AND "
```
General rule: Prefer parameterized queries over query string formatting. Usually faster and MUCH safer -> SQL injection. Doesn't matter here, but we should get rid of this naive query parameterization style that can be found in many places in the Mixxx code base.

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -414,6 +414,37 @@ void PlaylistDAO::removeHiddenTracks(const int playlistId) {
+        LOG_FAILED_QUERY(query);
+        return;
+    }
+    query.setForwardOnly(true);
```
You don't need this if you use our FwdSqlQuery. Usage examples can be found in CrateStorage.

uklotzde commented on src/library/dao/playlistdao.cpp:
```diff
@@ -414,6 +414,37 @@ void PlaylistDAO::removeHiddenTracks(const int playlistId) {
 }
 
+void PlaylistDAO::removeEmptyPlaylists(const int hidden) {
+    ScopedTransaction transaction(m_database);
```
We don't really need a transaction here for safety reasons, because each playlist can be deleted independently. It also doesn't work as expected, because SQLite doesn't allow nested transactions and deletePlaylist() internally tries to open another transaction that is committed immediately!

For the future: Transactions should be managed by a controller that coordinates operations on multiple DAOs. If operations in the DAO require to be executed within a transaction (because multiple, dependent SQL commands are executed subsequently) then a transaction should be passed as an additional parameter and the operation should check that this transaction is active before performing any SQL commands.

uklotzde replied:
Ups. I noticed that some operations in CrateStorage would actually need such a transaction parameter, because multiple mutating SQL commands are executed in a row. We should fix that, after the work of @gramanas on nested crates has been merged.

Be-ing commented:
If you're interested in the usability of the library, take a look at the library redesign branch. You can follow the discussion on #1117 and #1304. Currently we do not have any developers actively putting much time into it, so any help would be appreciated. :)

poelzi commented:
I think I addressed everything and this is ready for merge.

uklotzde commented:
@Be-ing LGTM. Approval from another reviewer?

@poelzi Did you sign the contributor agreement? I already asked Daniel how to check this in the context of your other PR.

uklotzde commented:
```
Debug [Main]: remove empty playlist: 49 
Debug [Main]: Committing transaction on "MIXXX-1" result: true 
Debug [Main]: PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  49 
Debug [Main]: PlaylistDAO::getHiddenType returns PLHT_UNKNOWN for playlistId  49 
Critical [Main]: DEBUG ASSERT: "feature() != nullptr" in function void TreeItem::insertChildren(QList<TreeItem*>&, int, int) at src/library/treeitem.cpp:90
```

uklotzde commented:
```
#5  0x00000000009e66d7 in mixxx_debug_assert(char const*, char const*, int, char const*) (function=<synthetic pointer>, line=90, file=0xfcb510 "src/library/treeitem.cpp", assertion=0xfcb560 "feature() != nullptr") at src/util/assert.h:11
#6  0x00000000009e66d7 in TreeItem::insertChildren(QList<TreeItem*>&, int, int) (this=this@entry=0x3a5dce0, children=..., row=row@entry=0, count=2634)
    at src/library/treeitem.cpp:90
#7  0x00000000009e73c2 in TreeItemModel::insertTreeItemRows(QList<TreeItem*>&, int, QModelIndex const&) (this=0x3a1bcb0, rows=..., position=0, parent=...)
    at src/library/treeitemmodel.cpp:176
#8  0x00000000008754c6 in BasePlaylistFeature::constructChildModel(int) (this=0x3a1bbf0, selected_id=<optimized out>)
    at src/library/baseplaylistfeature.cpp:646
#9  0x00000000009cdf8d in SetlogFeature::slotPlaylistTableChanged(int) (this=0x3---Type <return> to continue, or q <return> to quit--- 
a1bbf0, playlistId=49) at src/library/setlogfeature.cpp:292
#10 0x0000000000982636 in BasePlaylistFeature::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_a=<optimized out>, _id=<optimized out>, _c=QMetaObject::InvokeMetaMethod, _o=<optimized out>)
    at lin64_build/library/moc_baseplaylistfeature.cc:97
#11 0x0000000000982636 in BasePlaylistFeature::qt_metacall(QMetaObject::Call, int, void**) (this=<optimized out>, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>) at lin64_build/library/moc_baseplaylistfeature.cc:153
#12 0x00007ffff453d92d in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib64/libQtCore.so.4
#13 0x000000000090e32f in PlaylistDAO::deleted(int) (this=this@entry=0x16f0238, _t1=<optimized out>, _t1@entry=49)
    at lin64_build/library/dao/moc_playlistdao.cc:124
#14 0x0000000000912aee in PlaylistDAO::deletePlaylist(int) (this=this@entry=0x16f0238, playlistId=playlistId@entry=49) at src/library/dao/playlistdao.cpp:199
#15 0x0000000000914e19 in PlaylistDAO::removeEmptyPlaylists(PlaylistDAO::HiddenType) (this=0x16f0238, hidden=hidden@entry=PlaylistDAO::PLHT_SET_LOG)
    at src/library/dao/playlistdao.cpp:445
#16 0x00000000009d0481 in SetlogFeature::SetlogFeature(QObject*, QSharedPointer<ConfigObject<ConfigValue> >, TrackCollection*) (this=0x3a1bbf0, parent=<optimized out>, pConfig=..., pTrackCollection=0x16f0220)
    at src/library/setlogfeature.cpp:24
#17 0x000000000096df94 in Library::Library(QObject*, QSharedPointer<ConfigObject---Type <return> to continue, or q <return> to quit---
<ConfigValue> >, std::shared_ptr<mixxx::DbConnectionPool>, PlayerManagerInterface*, RecordingManager*) (this=0x3697d20, parent=<optimized out>, pConfig=..., pDbConnectionPool=..., pPlayerManager=<optimized out>, pRecordingManager=<optimized out>, __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at src/library/library.cpp:126
#18 0x0000000000a0a5a4 in MixxxMainWindow::initialize(QApplication*, CmdlineArgs const&) (this=this@entry=0x7fffffffd9e0, pApp=pApp@entry=0x7fffffffd9b0, args=...) at src/mixxx.cpp:288
#19 0x0000000000a0ce55 in MixxxMainWindow::MixxxMainWindow(QApplication*, CmdlineArgs const&) (this=0x7fffffffd9e0, pApp=0x7fffffffd9b0, args=...)
    at src/mixxx.cpp:158
#20 0x00000000004677ae in (anonymous namespace)::runMixxx (args=..., app=0x7fffffffd9b0) at src/main.cpp:43
```

poelzi committed:
Fix assert.
The deletion happend when parts of mixxx where initialized but not all signals connected.
Doing the cleanup earlier prevents the assert.

uklotzde commented on src/library/setlogfeature.cpp:
```diff
@@ -17,11 +17,11 @@ SetlogFeature::SetlogFeature(QObject* parent,
         : BasePlaylistFeature(parent, pConfig, pTrackCollection, "SETLOGHOME"),
           m_playlistId(-1) {
+    // we cleanup old history playlists that don't have any tracks
+    m_playlistDao.removeEmptyPlaylists(PlaylistDAO::HiddenType::PLHT_SET_LOG);
```
It would be a good idea to leave a note in the code. It is not obvious that the order is crucial here, e.g. something like // Cleanup database BEFORE creating and connecting the corresponding table model

poelzi committed:
add comment to cleanup order

uklotzde commented:
On the first start Mixxx removed 100 or even more of my empty playlists, that took about 20 sec. with debug logging enabled.

After starting up a library rescan started in the background (unintentionally, no changes on the file system) and I clicked on two or three of the remaining history playlists just to check that they are not empty. When selecting such a non-empty history playlists Mixxx stopped with another debug assertion:
```
Critical [Main]: DEBUG ASSERT: "trackIdToRows.size() == rowInfos.size()" in function virtual void BaseSqlTableModel::select() at src/library/basesqltablemodel.cpp:351

#0  0x00007ffff1ce190b in raise () at /usr/lib64/libpthread.so.0
#1  0x0000000000c2a3e4 in mixxx::(anonymous namespace)::MessageHandler(QtMsgType, char const*) (type=<optimized out>, input=<optimized out>)
    at src/util/logging.cpp:154
#2  0x00007ffff4412666 in qt_message_output(QtMsgType, char const*) ()
    at /usr/lib64/libQtCore.so.4
#3  0x00007ffff4412851 in qt_message(QtMsgType, char const*, __va_list_tag*) ()
    at /usr/lib64/libQtCore.so.4
#4  0x00007ffff4412b81 in qCritical(char const*, ...) ()
    at /usr/lib64/libQtCore.so.4
#5  0x000000000087529b in mixxx_debug_assert(char const*, char const*, int, char const*) (function=<synthetic pointer>, line=351, file=0xf90610 "src/library/basesqltablemodel.cpp", assertion=0xf90720 "trackIdToRows.size() == rowInfos.size()") at src/util/assert.h:11
#6  0x000000000087529b in BaseSqlTableModel::select() (this=0x3a51120)
    at src/library/basesqltablemodel.cpp:351
#7  0x0000000000d3617d in WTrackTableView::doSortByColumn(int) (this=this@entry=0x56513cf0, headerSection=1) at src/widget/wtracktableview.cpp:1529
#8  0x0000000000d36a52 in WTrackTableView::loadTrackModel(QAbstractItemModel*) (this=this@entry=0x56513cf0, model=0x3a51120)
    at src/widget/wtracktableview.cpp:225
#9  0x0000000000cb962d in WTrackTableView::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=0x56513cf0, _c=<optimized out>, _id=0, _a=0x7fffffff---Type <return> to continue, or q <return> to quit---
c740) at lin64_build/widget/moc_wtracktableview.cc:111
#10 0x00007ffff453d92d in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib64/libQtCore.so.4
#11 0x000000000097f7bd in Library::showTrackModel(QAbstractItemModel*) (this=this@entry=0x367e120, _t1=<optimized out>, _t1@entry=0x3a51120)
    at lin64_build/library/moc_library.cc:186
#12 0x0000000000963173 in Library::slotShowTrackModel(QAbstractItemModel*) (this=this@entry=0x367e120, model=0x3a51120) at src/library/library.cpp:293
#13 0x000000000097fd20 in Library::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=0x367e120, _c=<optimized out>, _id=15, _a=0x7fffffffc8f0)
    at lin64_build/library/moc_library.cc:120
#14 0x00007ffff453d92d in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib64/libQtCore.so.4
#15 0x00000000009812fd in LibraryFeature::showTrackModel(QAbstractItemModel*) (this=this@entry=0x39e2470, _t1=<optimized out>)
    at lin64_build/library/moc_libraryfeature.cc:145
#16 0x0000000000866cbc in BasePlaylistFeature::activateChild(QModelIndex const&) (this=0x39e2470, index=...) at src/library/baseplaylistfeature.cpp:144
#17 0x000000000098367d in SidebarModel::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=0x3680640, _c=<optimized out>, _id=1, _a=0x7fffffffca40) at lin64_build/library/moc_sidebarmodel.cc:79
#18 0x00007ffff453d92d in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib64/libQtCore.so.4
```

Any idea what might have caused this inconsistency? I haven't checked yet.

poelzi commented:
I was not able to reproduce your debug asserts. I added hundrets of empty playlists and tried different startups and tests and never got a assert. Maybe it's a timing issue.
@uklotzde can you try this patch ?

```
diff --git a/src/library/setlogfeature.cpp b/src/library/setlogfeature.cpp                                                                  
index 1d889d9f0..a05ee1e6b 100644                                                                                                           
--- a/src/library/setlogfeature.cpp
+++ b/src/library/setlogfeature.cpp
@@ -21,6 +21,8 @@ SetlogFeature::SetlogFeature(QObject* parent,
     // it's important that we do this before we bind the new model, otherwise
     // it will have reference to soon deleted playlists
     m_playlistDao.removeEmptyPlaylists(PlaylistDAO::HiddenType::PLHT_SET_LOG);
+    // cleanup internal state
+    buildPlaylistList();
     m_pPlaylistTableModel = new PlaylistTableModel(this, pTrackCollection,
                                                    "mixxx.db.model.setlog",
                                                    true); //show all tracks
```

poelzi commented:
@uklotzde are you sure you are running this branch. I don't even find the debug assert you are referring to and the line numbers are all different. "trackIdToRows.size() == rowInfos.size()" does not exist here :)

poelzi commented:
Ahh, the bug got introduced with commit a9a5dbb0612fb7905a755d7122bba81b46bbe639. It starts to assume that all rows returned by select are unique tracks which is of course wrong. A playlist may contain 2 times the same track and then the assert fails. Has nothing todo with the cleanup but every playlist with two times the same track will crash on debug assert.

uklotzde commented:
@poelzi Just checked, the assertion is already in master. I've added this check myself just recently when doing some minor cleanup in BaseSqlTableModel. It has not failed until now. Need to check again if this assumption is correct or not!

I also have a backup and consider to repeat the history cleanup. Please be patient ;)

uklotzde commented:
👍 My bad 🙄  Would you please remove it or fix the condition? A short comment that explains this fact might also help.

uklotzde commented:
@poelzi Please leave the wrong debug assertion as is, because it is unrelated. I will post a PR with a fix and some notes in the code.

uklotzde commented:
I will retest you branch with a database file from my backup, just to be sure ;)

uklotzde commented:
No issues after I restored my database from a backup and retested the cleanup. The cleanup takes some time, because deleting individual playlists triggers subsequent database queries.

For crates both cleanup and consistency checks are executed when connecting to the database before any slots are connected. Actually the database layer does not even use signals and slots anymore. Signals and slots are over- and even misused in Mixxx. We are doomed if we don't clean up that mess. They will become relevant again when implementing asynchronous database access with separate query thread(s).

Any objections? Otherwise I would merge this PR.

uklotzde commented:
I'm still a bit worried about the extended startup time that users will experience after upgrading to Mixxx 2.1. We don't know how many empty playlists will be deleted and how long this will take in the worst case. Leaving a user with a frozen startup screen would be a bad first impression. I think we are talking about less than a minute, though.

Any thoughts or personal test results?

esbrandt commented:
> I'm still a bit worried about the extended startup time that users will experience after upgrading to Mixxx 2.1.
...

Agree, this would be unexpected. Users might think the application hangs, and do a force quit, with unwanted consequences.

Can´t we simply do the cleanup on first shutdown?
Displaying a small text similar to
``After upgrading to Mixxx 2.1, the application will now clean all empty history playlists on shutdown. This can take several minutes. Please be patient.`` 

The regular user will probably not have hundreds of empty playlists like a dev has, so clean-up time is expected to be significant faster.

uklotzde commented:
We shouldn't bother the user if there is a technically superior alternative.

**1.** Add a new function to PlaylistDAO:
```
private:
static void cleanupDatabase(QSqlDatabase sqlDatabase);
```
Both _private_ and _static_, because this function must not be invoked after PlaylistDAO has been initialized and not arbitrarily by anyone (at least not until we have a common concept for database consistency checks and cleanup).

**2.** Execute the following SQL commands in the implementation:
- Delete all empty playlist from table 'Playlists'
- Delete all orphan playlist tracks from table 'PlaylistTracks' without a corresponding playlist

The 2nd step will also cleanup entries that have been left behind from failed or aborted database operations. It is crucial that we only delete entries of non-existent playlists here! Deleting entries from existent playlists would require renumbering.

Transactions are not necessary, because it doesn't harm if we abort after the first command. We can repeat those 2 commands as often as we want, in conjunction they are _eventually idempotent_. It is sufficient to execute them upon every startup once.

**3.** Invoke cleanupDatabase() in PlaylistDAO::initialize(), before populating the membership cache.

Sorry, I should have proposed that solution earlier. I was not aware how massive the side effects actually are until I tried this at home.

See also: CrateStorage::repairDatabase()
We need to consolidate both approaches at some point.

uklotzde reviewed (changes requested):
Improve execution time by cleaning up Playlists and PlaylistTracks with separate SQL commands when initializing PlaylistDAO.

poelzi commented:
@uklotzde you are right. I just thought about the same thing, I will look at it, maybe there can be some nice generic solution for all storage backends.

Be-ing commented:
@poelzi a merge conflict with  src/library/setlogfeature.cpp has appeared.

Xerus2000 commented:
Hello, could you please finally merge this?

poelzi commented:
superseeded by #2996
